baseapp/msg_service_router.go

```diff
@@ -45,21 +45,20 @@ func (msr *MsgServiceRouter) RegisterService(sd *grpc.ServiceDesc, handler inter
 		fqMethod := fmt.Sprintf("/%s/%s", sd.ServiceName, method.MethodName)
 		methodHandler := method.Handler
 
-		// NOTE: This is how we pull the concrete request type for each handler for registering in the InterfaceRegistry.
-		// This approach is maybe a bit hacky, but less hacky than reflecting on the handler object itself.
-		// We use a no-op interceptor to avoid actually calling into the handler itself.
-		_, _ = methodHandler(nil, context.Background(), func(i interface{}) error {
-			msg, ok := i.(proto.Message)
-			if !ok {
-				// We panic here because there is no other alternative and the app cannot be initialized correctly
-				// this should only happen if there is a problem with code generation in which case the app won't
-				// work correctly anyway.
-				panic(fmt.Errorf("can't register request type %T for service method %s", i, fqMethod))
-			}
-
-			msr.interfaceRegistry.RegisterCustomTypeURL((*sdk.MsgRequest)(nil), fqMethod, msg)
-			return nil
-		}, noopInterceptor)
+		// Check that the service Msg fully-qualified method name has already
+		// been registered (via RegisterInterfaces).
+		serviceMsg, err := msr.interfaceRegistry.Resolve(fqMethod)
+		if err != nil || serviceMsg == nil {
+			panic(
```

Comment:
@amaurymartiny do you think it's preferable to panic, or just write a warning to stdout?

Reply:
I think panicking is alright, since this is called at startup and gives immediate dev feedback that something is wrong in the code.

I guess the reverse question would be: why shouldn't we panic? Would we allow a case for app developers to:
```go
app.RegisterService()
// let the app developer do something else
// ...
// 
app.mm.RegisterInterfaces()
```
? Do you have some ideas in mind?

Reply:
We should avoid adding `panic` whenever possible. I would say the only exceptions are if this would require bigger refactor.

Panic is bad because:
* it will hide the error structure
* it's harder to test
* doesn't show to the user that the method can panic. On the other hand, if a function returns error, it's clear to the user that something can go wrong.

In general, in Go we discourage use of `panic`. It's  use is rather limited to some hacks or frameworks.

Reply:
> I guess the reverse question would be: why should we allow the app to continue running? 

A user should make that decision. Propagating an error is the right way. We may continue running the app without that service.

Reply:
Please see also this comment for more thoughts: https://github.com/cosmos/cosmos-sdk/pull/7674#discussion_r512119900

Reply:
@robert-zaremba I +1 the effort to never panic. I wrote somewhere that my ideal SDK would be one with 0 panics, would propagate all app-related errors to TM (with some error spec), and only TM can decide to panic or not.

Circling back to the code. I'm implementing `gogogrpc.Server` here, so cannot return an error. Either we panic asap, or we log a warning.

Reply:
Exactly. Library should never panic, unless it is using it (eg some testing frameworks). If a library doesn't handle it, and leaves it to a user, then it's bad - errors are much better.

Reply:
So, let's ask a question about [this](https://github.com/cosmos/cosmos-sdk/pull/7671/files/87102598510b07a38f2587dd61152cdb98cfa6ff#r512126290):
> A user should make that decision. Propagating an error is the right way. We may continue running the app without that service.

If yes then we have a design problem. Otherwise, let's just panic :(

Reply:
.. and add a proper comment

Reply:
> A user should make that decision. Propagating an error is the right way. We may continue running the app without that service.

Yes, that was my question. Since we cannot propagate errors, should we:
- panic asap to tell the app developer he did something risky?
- or log an error, and let the user handle its app (with a high risk that it panics down the road)?

Reply:
Added comment 👍

Reply:
Please add also a comment that the function panics to the function doc-comment:
```
...
// Function PANICs when ....
func (msr *MsgServiceRouter) RegisterService(...
```

Reply:
What about just a stdout warning?